File: openbis_api/source/java/ch/ethz/sis/openbis/generic/asapi/v3/dto/common/update/RelationshipUpdate.java
/*
 * Copyright 2020 ETH Zuerich, SIS
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ch.ethz.sis.openbis.generic.asapi.v3.dto.common.update;

import java.io.Serializable;

import com.fasterxml.jackson.annotation.JsonIgnore;

import ch.ethz.sis.openbis.generic.asapi.v3.dto.common.ObjectToString;
import ch.ethz.sis.openbis.generic.asapi.v3.dto.common.Relationship;
import ch.systemsx.cisd.base.annotation.JsonObject;

/**
 * @author Franz-Josef Elmer
 */
@JsonObject("as.dto.common.update.RelationshipUpdate")
public class RelationshipUpdate implements Serializable
{
    private static final long serialVersionUID = 1L;

    private ListUpdateMapValues childAnnotations = new ListUpdateMapValues();
    private ListUpdateMapValues parentAnnotations = new ListUpdateMapValues();

    @JsonIgnore
    public RelationshipUpdate addChildAnnotation(String key, String value)
    {
        childAnnotations.put(key, value);
        return this;
    }

    @JsonIgnore
    public RelationshipUpdate removeChildAnnotations(String... keys)
    {
        childAnnotations.remove(keys);
        return this;
    }

    @JsonIgnore
    public RelationshipUpdate addParentAnnotation(String key, String value)
    {
        parentAnnotations.put(key, value);
        return this;
    }

    @JsonIgnore
    public RelationshipUpdate removeParentAnnotations(String... keys)
    {
        parentAnnotations.remove(keys);
        return this;
    }

    @SuppressWarnings("unchecked")
    @JsonIgnore
    public void setRelationship(Relationship relationship)
    {
        childAnnotations.set(relationship.getChildAnnotations());
        parentAnnotations.set(relationship.getParentAnnotations());
    }
    
    public ListUpdateMapValues getChildAnnotations()
    {
        return childAnnotations;
    }

    public void setChildAnnotations(ListUpdateMapValues childAnnotations)
    {
        this.childAnnotations = childAnnotations;
    }

    public ListUpdateMapValues getParentAnnotations()
    {
        return parentAnnotations;
    }

    public void setParentAnnotations(ListUpdateMapValues parentAnnotations)
    {
        this.parentAnnotations = parentAnnotations;
    }

    @Override
    public String toString()
    {
        return new ObjectToString(this).append("child annotations update", childAnnotations)
                .append("parent annotations update", parentAnnotations).toString();
    }
}

File: openbis/source/java/ch/systemsx/cisd/openbis/public/resources/api/v3/as/dto/common/update/RelationshipUpdate.js
define([ "stjs", "as/dto/common/update/ListUpdateMapValues" ], function(stjs, ListUpdateMapValues) {
	var RelationshipUpdate = function() {
		this.childAnnotations = new ListUpdateMapValues();
		this.parentAnnotations = new ListUpdateMapValues();
	};
	stjs.extend(RelationshipUpdate, null, [], function(constructor, prototype) {
		prototype['@type'] = 'as.dto.common.update.RelationshipUpdate';
		constructor.serialVersionUID = 1;
		prototype.childAnnotations = null;
		prototype.parentAnnotations = null;
		
		prototype.getChildAnnotations = function() {
			return this.childAnnotations;
		};
		prototype.setChildAnnotations = function(childAnnotations) {
			this.childAnnotations = childAnnotations;
		};
		prototype.addChildAnnotation = function(key, value) {
			this.childAnnotations.put(key, value);
			return this;
		};
		prototype.removeChildAnnotations = function(keys) {
			this.childAnnotations.remove(keys)
			return this;
		};
		prototype.getParentAnnotations = function() {
			return this.parentAnnotations;
		};
		prototype.setParentAnnotations = function(parentAnnotations) {
			this.parentAnnotations = parentAnnotations;
		};
		prototype.addParentAnnotation = function(key, value) {
			this.parentAnnotations.put(key, value);
			return this;
		};
		prototype.removeParentAnnotations = function(keys) {
			this.parentAnnotations.remove(keys)
			return this;
		};
		prototype.setRelationship = function(relationship) {
			this.childAnnotations.set(relationship.getChildAnnotations());
			this.parentAnnotations.set(relationship.getParentAnnotations());
		};
	}, {
		childAnnotations : "ListUpdateMapValues",
		parentAnnotations : "ListUpdateMapValues"
	});
	return RelationshipUpdate;
})
